packages/spacecat-shared-ims-client/src/clients/ims-client.js

```diff
@@ -0,0 +1,252 @@
+        ...(isObject(body) ? { body: createFormData(body) } : {}),
+      },
+    );
+  }
+
+  async #getImsOrgDetails(imsOrgId) {
+    const orgDetailsResponse = await this.#imsApiCall(
+      getImsOrgsApiPath(imsOrgId),
+      { client_id: this.config.clientId },
+    );
+
+    if (!orgDetailsResponse.ok) {
+      throw new Error(`IMS getImsOrgDetails request failed with status: ${orgDetailsResponse.status}`);
+    }
+
+    return orgDetailsResponse.json();
+  }
+
+  async #getProductContextByImsOrgId(imsOrgId) {
+    const { orgId, authSource } = extractIdAndAuthSource(imsOrgId);
+
+    const pcResponse = await this.#imsApiCall(
+      IMS_PRODUCT_CONTEXT_BY_ORG_ENDPOINT,
+      {},
+      {
+        org_id: orgId,
+        auth_src: authSource,
+        client_id: this.config.clientId,
+      },
+      true,
+    );
+
+    if (!pcResponse.ok) {
+      throw new Error(`IMS getProductContextsByImsOrgId request failed with status: ${pcResponse.status}`);
+    }
+
+    const pcData = await pcResponse.json();
+
+    // Always return the first product context item
+    return pcData?.productContexts?.[0];
+  }
+
+  async #getUsersByImsGroupId(imsOrgId, groupId) {
+    // This endpoint is paginated, but the default page limit is 50 entries — more than enough
+    // for our use case
+    const groupResponse = await this.#imsApiCall(
+      getGroupMembersEndpoint(imsOrgId, groupId),
+      { client_id: this.config.clientId },
+    );
+
+    if (!groupResponse.ok) {
+      throw new Error(`IMS getUsersByImsGroupId request failed with status: ${groupResponse.status}`);
+    }
+
+    const group = await groupResponse.json();
+
+    return group.items || [];
+  }
+
+  async #getUsersInAdminGroup(imsOrgId, groups) {
+    if (!Array.isArray(groups)) {
+      return [];
+    }
+
+    // Store users by their email address initially to de-dupe the entries
+    const users = {};
+    for (const group of groups) {
+      // Only process Administrators groups
+      if (group?.role === 'GRP_ADMIN') {
+        // eslint-disable-next-line no-await-in-loop
+        const groupUsers = await this.#getUsersByImsGroupId(imsOrgId, group?.ident);
+        for (const user of groupUsers) {
+          // Fallback to username if email is not set
+          const newUser = { ...user };
+          if (!hasText(newUser.email)) {
+            newUser.email = newUser.username;
+          }
+          if (emailAddressIsAllowed(newUser.email)) {
+            // Reduce fields in user object to those we need
+            users[newUser.email] = {
+              email: newUser.email,
+              firstName: newUser.firstName,
+              lastName: newUser.lastName,
+            };
+          }
+        }
+      }
+    }
+
+    // Transform the object "map" back to a de-duped array
+    return Object.keys(users).map((email) => users[email]);
+  }
+
+  async getServiceAccessToken() {
+    if (hasText(this.serviceAccessToken)) {
+      return this.serviceAccessToken;
```

Comment:
heads-up: this function returns a string here but down below it returns an object